_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,387 @@
+---
+layout: blog-detail
+post-type: blog
+by: Vincenzo Bazzucchi, Scala Center
+title: Flexible and safe tuples in Scala 3
+---
+
+# Flexible and safe tuples in Scala 3
+
+Tuples are completely revisited in Scala 3.
+They are more **flexible**, more dynamic and support a **new range of operations**.
+This is enabled by new and powerful features in the Scala 3 compiler.
+
+In this post we will explore the new capabilities of tuples before
+looking under the hood to learn how the improvements in the Scala 3
+compiler enable to implement type safe operations on tuples.
+
+# The basics: what are tuples?
+
+In the Python programming language, tuples are a simple concept:
+they are immutable collections of objects. As such, they are opposed
+to lists, which are mutable.
+
+In Scala both `List`s and tuples are immutable, so why do we care
+about tuples?
+
+Scala being a statically typed programming language, the difference between
+list and tuples is in the type. Lists are *homogeneous* collections while
+tuples are *heterogeneous*. In simpler terms, a tuple collects items maintaining
+the type of each element, while a list collects objects retaining a common type
+for all the elements.
+
+This is better explained with an example:
+```scala
+scala> List(1, "2", 3.0, List(4))
+val res0: List[Any] = List(1, 2, 3.0, List(4))
+```
+We see that the compiler tries to infer a common supertype for the elements of the list,
+in this case `Any`.
+
+If do the same with tuples, the elements maintain their individual and specific type:
+```scala
+scala> (1, "2", 3.0, List(4))
+val res0: (Int, String, Double, List[Int]) = (1, 2, 3.0, List(4))
+```
+This behavior is desirable in many cases, for example when
+we want a function to return two or more values or when we want
+to use types to decide which transformation to apply to each
+value.
+
+# How do tuples evolve in Scala 3?
+
+## Size limit
+
+Probably the most well known limitation of tuples in Scala 2 was the
+restriction to 22 for the number of elements.
+
+```scala
+scala> (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23)
+    error: tuples may not have more than 22 elements, but 23 given
+```
+
+In Scala 3 the previous tuple is perfectly legal.
+
+## Element accessor
+
+The only way to retrieve an element of a tuple in Scala 2 was to
+use the (1-based) `._i` attribute. For example:
+
+```scala
+("First", "Second")._2 // "Second"
+```
+
+In Scala 3, we can use the `apply` method with a 0-based argument:
+
+```scala
+("First", "Second")(1) // "Second"
+```
+
+As most of indexes are 0 based in Scala, this brings more consistency
+to codebases. It also provides more flexibility. We can, for example,
+*iterate* on any tuple to print each element on a line:
+
+```scala
+val someStuff = (1, "2", 3.0, List(4))
+for (i <- 0 until someStuff.size)
+  println(someStuff(i))
+```
+
```

Comment:
I'm  a bit uncomfortable with this example because it seems like something one shouldn't do. What is the _purpose_ of allowing indexed access like this? Is the main purpose consistency? Or is the main purpose that we actually expect such looping-over-a-tuple to be common? I doubt it will be common, and so using it too prominently in an example runs the risk of confusing readers about what tuples are for.

If iterating over a tuple is uncommon but something one might _occasionally_ want to do, then I think it would be good to clearly signal that.